### PR TITLE
key: Improve unit tests

### PR DESCRIPTION
**Pull request by tcharding**
Improve the unit tests in the `key` module by doing:

- Remove a bunch of "rand-std" (use hard coded keys instead)
- Split big tests into smaller appropriately named tests
- Move things around so we can better see whats being tested

### CODE REVIEW

**Comment by tcharding** [comment]
I got this idea for a post you made someplace or other about not having tests when "rand-std" feature was disabled. If this PR is good we can probably mimic it for a bunch of other modules.

File: src/key.rs [hunk 1]
```diff
@@ -1560,10 +1560,42 @@ mod test {
...
+    fn keys() -> (SecretKey, PublicKey, KeyPair, XOnlyPublicKey) {
+        let secp = Secp256k1::new();
+
+        #[rustfmt::skip]
+        static SK_BYTES: [u8; 32] = [
+            0x01, 0x01, 0x01, 0x01, 0x01, 0x01, 0x01, 0x01,
+            0x00, 0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07,
+            0xff, 0xff, 0x00, 0x00, 0xff, 0xff, 0x00, 0x00,
+            0x63, 0x63, 0x63, 0x63, 0x63, 0x63, 0x63, 0x63,
+        ];
```
Comment (Kixunil) [inline src/key.rs:1577]: Could also use `hex_lit::hex!`

File: src/key.rs [hunk 2]
```diff
@@ -1560,10 +1560,42 @@ mod test {
...
+        static PK_BYTES: [u8; 32] = [
+            0x18, 0x84, 0x57, 0x81, 0xf6, 0x31, 0xc4, 0x8f,
+            0x1c, 0x97, 0x09, 0xe2, 0x30, 0x92, 0x06, 0x7d,
+            0x06, 0x83, 0x7f, 0x30, 0xaa, 0x0c, 0xd0, 0x54,
+            0x4a, 0xc8, 0x87, 0xfe, 0x91, 0xdd, 0xd1, 0x66
+        ];
+
+        let mut pk_bytes = [0u8; 33];
+        pk_bytes[0] = 0x02; // Use positive Y co-ordinate.
+        pk_bytes[1..].clone_from_slice(&PK_BYTES);
```
Comment (Kixunil) [inline src/key.rs:1589]: I'm surprised clippy doesn't scream about `copy`.

File: src/key.rs [hunk 3]
```diff
@@ -1652,184 +1655,174 @@ mod test {
...
+    fn invalid_secret_key_one_too_long_slice_compressed() {
         assert_eq!(
-            PublicKey::from_slice(&[0; constants::UNCOMPRESSED_PUBLIC_KEY_SIZE - 1]),
-            Err(InvalidPublicKey)
-        );
-        assert_eq!(
-            PublicKey::from_slice(&[0; constants::UNCOMPRESSED_PUBLIC_KEY_SIZE + 1]),
-            Err(InvalidPublicKey)
+            SecretKey::from_slice(&[0; constants::SECRET_KEY_SIZE + 1]),
+            Err(InvalidSecretKey)
```
Comment (Kixunil) [inline src/key.rs:1669]: This test isn't that great since all-zero keys are also invalid.

File: src/key.rs [hunk 4]
```diff
@@ -1850,121 +1843,176 @@ mod test {
...
+    fn erased_keypair_is_valid() {
+        let s = Secp256k1::new();
+        let kp = KeyPair::from_seckey_slice(&s, &[1u8; constants::SECRET_KEY_SIZE])
+            .expect("valid secret key");
+        let mut kp2 = kp;
+        kp2.non_secure_erase();
+        assert!(kp.eq_fast_unstable(&kp2));
+    }
+
+    // FIXME(Tobin): I don't know what this is testing? There is no assertion or panic?
```
Comment (Kixunil) [inline src/key.rs:1866]: I believe this tests that if an RNG generates an out-of-range key it is called again instead of panicking.

**Comment by apoelstra** [comment]
Maybe we should instead have a `TestRng` object that internally uses `thread_rng` when we have rand-std but otherwise uses a LCG or something (or even uses a custom super-biased RNG that likes te produce extreme values). Then our tests can just call this object.

File: src/key.rs (hunk 3 above)
Comment (tcharding) [reply src/key.rs:1669]: Nice observation, I'll improve it.

File: src/key.rs (hunk 4 above)
Comment (tcharding) [reply src/key.rs:1866]: oooh, thanks man.

File: src/key.rs [hunk 5]
```diff
@@ -1850,121 +1843,176 @@ mod test {
...
     #[test]
-    #[cfg(feature = "rand-std")]
+    #[cfg(all(not(fuzzing), feature = "alloc"))]
     fn tweak_mul_arbitrary_data() {
-        let s = Secp256k1::new();
-
-        let (sk, pk) = s.generate_keypair(&mut rand::thread_rng());
-        assert_eq!(PublicKey::from_secret_key(&s, &sk), pk); // Sanity check.
-
-        // TODO: This would be better tested with a _lot_ of different tweaks.
```
[the remark this answers is not in the document]
Comment (tcharding) [reply src/key.rs:1927]: Implying that running the tests multiple times improves coverage? The test space (a random `Scalar`) is so large that no amount of test re-runs in any way improves coverage so I think the use of `random` gives a false sense of coverage. I've not worked out how to test this (and the other tests still "rand-std") though, I tried playing with kani yesterday without success. Open to suggestions but I probably just need to play with it some more.

**Comment by tcharding** [comment]
> or even uses a custom super-biased RNG that likes te produce extreme values

That's a cool idea.

**Comment by tcharding** [comment]
Thanks for the reviews, I didn't respond to all your comments @Kixunil but I read them and I'll come back to this PR when I'm in a "psyched for testing" mood.

File: src/key.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (Kixunil) [reply src/key.rs:1927]: Good point. Maybe just test a few obvious things that are generally edge-cases - all zeroes, max...

**Comment by tcharding** [comment]
Seems "psyched for testing" is taking a while to materialise, converting to draft.

**Comment by tcharding** [comment]
Closing as stale during migration to Forgejo.
